gh-111201: A new Python REPL

pablogsal opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-111201 -->
* Issue: gh-111201
<!-- /gh-issue-number -->

pablogsal committed:
gh-111201: A new Python REPL

vstinner commented on Lib/_pyrepl/__main__.py:
```diff
@@ -0,0 +1,78 @@
+
+        run_interactive = run_multiline_interactive_console
+    # except ImportError:
+    #     pass
```
Is it dead code?

vstinner commented on Lib/_pyrepl/__main__.py:
```diff
@@ -0,0 +1,78 @@
+    if os.getenv("PYTHONSTARTUP"):
+        exec(
+            compile(
+                open(os.getenv("PYTHONSTARTUP")).read(),
```
I would prefer to use with to avoid ResourceWarning. tokenize.open() may be better to use the right encoding.

vstinner commented on Lib/_pyrepl/curses.py:
```diff
@@ -0,0 +1,36 @@
+# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+# If we are running on top of pypy, we import only _minimal_curses.
+# Don't try to fall back to _curses, because that's going to use cffi
```
Are these comments relevant for CPython?

vstinner commented on Python/pythonrun.c:
```diff
@@ -83,8 +83,6 @@ _PyRun_AnyFileObject(FILE *fp, PyObject *filename, int closeit,
 }
 
-
-/* Parse input from a file and execute it */
```
This comment is no longer relevant?

vstinner commented on Lib/_pyrepl/__main__.py:
```diff
@@ -0,0 +1,78 @@
+import os
+import sys
+
+irc_header = "And now for something completely different"
```
Is it an easter egg or is it used somewhere?

vstinner reviewed (commented):
You add many env vars. Can you document them in Doc/using/cmdline.rst?

Usually, env var names start with PYTHON. Should the variables be renamed?

Also, Python ignores env var starting with PYTHON if sys.flags.ignore_environment.

vstinner commented on Lib/_pyrepl/_minimal_curses.py:
```diff
@@ -0,0 +1,68 @@
+manually.
+
+Note that there is also a built-in module _minimal_curses which will
+hide this one if compiled in.
```
Is it part of this PR? If not, do you consider adding it or not?

JelleZijlstra replied:
Yes, sounds like this comment was carried over from pypy but doesn't apply here.

pablogsal committed:
Clean some files

pablogsal committed:
Fix history across sessions

pablogsal committed:
Fore fixes to history

pablogsal committed:
Lukasz was wrong all along

pablogsal committed:
Implement REPL commands and remove F1 for help (curses shows that scrambled)

pablogsal committed:
Restore F1 for help

pablogsal committed:
Implement colored prompt

pablogsal committed:
WIP for paste mode

ambv committed:
Only execute PYTHONSTARTUP when asked (we do it in main.c anyway)

pablogsal committed:
Fixed coloring

pablogsal committed:
Do not run commands if they are shadowed

ambv committed:
Move pager routines to _pyrepl, so we can use it in the REPL without importing too much

ambv committed:
Don't assume stupid paths or you will be tested against your stupid assumptions

ambv committed:
<F2> shows history so you can copy it, <F3> is paste mode

pablogsal committed:
Do not include REPL commands in history

ambv committed:
Add Blurb

ambv committed:
Reintroduce old names to fix pydoc tests

pablogsal committed:
Add some tests

pablogsal committed:
Add moar tests

pablogsal committed:
Add moar tests

pablogsal committed:
Cleanup and refactor

pablogsal committed:
Refactor unix_console.py

pablogsal committed:
Refactor unix_console.py more

pablogsal committed:
Refactor unix_console.py more

pablogsal committed:
Moar tests

ambv committed:
Fix test___all__

ambv committed:
Fix test_traceback

pablogsal committed:
Fix weird r-search scrambled text due to color codes

pablogsal committed:
Fix weird r-search scrambled text due to color codes better

pablogsal committed:
Disgusting fix for help state restoration

mgmacias95 committed:
Test pasting with/without paste mode

mgmacias95 committed:
refactor events

pablogsal committed:
Merge pull request #71 from mgmacias95/pyrepl

Test pasting with/without paste mode

ambv committed:
help() uses its own history, doesn't pollute main history

ambv committed:
Add typing to reader.py

ambv committed:
Add types to pager.py

ambv committed:
Add types to input.py, console.pu, and commands.py

lysnikolaou committed:
Refactor termios stuff in unix console

pablogsal committed:
Merge pull request #72 from lysnikolaou/pyrepl-refactor-term-unix-console

Refactor termios stuff in unix console

pablogsal committed:
Revert "Refactor termios stuff in unix console"

This reverts commit 2a7e81d3d772a23fe9803526916a56eb4fe0788b.

lysnikolaou committed:
Fix cursor position for double-width characters

lysnikolaou committed:
Add _pyrepl to installed Lib subdirs

pablogsal committed:
Merge pull request #74 from lysnikolaou/pyrepl-add-module-install

lysnikolaou committed:
Add tests

pablogsal committed:
Update Lib/test/test_pyrepl.py

pablogsal committed:
Merge pull request #73 from lysnikolaou/pyrepl-double-width-char-cursor

Fix cursor position for double-width characters

pablogsal committed:
Implement better fallback with PYTHON_OLD_REPL

pablogsal committed:
Cache failures in pyrepl

lysnikolaou committed:
Stay at eol when moving up/down

lysnikolaou committed:
Fix linter

pablogsal committed:
fix early errors

pablogsal committed:
Fix CI

lysnikolaou committed:
Fix paste mode when there are empty line in the middle

Paste mode needs to be exited with F3 after pasting now.

Co-authored-by: Pablo Galindo <pablogsal@gmail.com>

lysnikolaou committed:
Fix disp_str for control characters

pablogsal committed:
Fix mac CI

pablogsal committed:
Merge pull request #77 from lysnikolaou/pyrepl-fix-disp-str-ctrl

Fix disp_str for control characters

pablogsal committed:
Merge pull request #76 from lysnikolaou/pyrepl-paste-mode-newlines-mid

Fix paste mode when there are empty line in the middle

pablogsal committed:
Merge pull request #75 from lysnikolaou/pyrepl-eol-move-up-down

Stay at eol when moving up/down

lysnikolaou committed:
Various fixes to handle wide characters correctly

Many of the fixes incorporated from https://github.com/pypy/pyrepl/pull/34/files

lysnikolaou committed:
Fix vertical navigation with wide characters

lysnikolaou committed:
Write more tests

lysnikolaou committed:
Fix backspace in second line

lysnikolaou committed:
Fixes for pos2xy & setpos_from_xy when wrapped line

lysnikolaou committed:
Fix linter & run black

pablogsal committed:
Update Lib/_pyrepl/reader.py

pablogsal committed:
Merge pull request #78 from lysnikolaou/pyrepl-various-fixed-wide-characters

Various fixes to handle wide characters correctly

ambv committed:
Add typing to consoles

ambv committed:
Merge branch 'main' into pyrepl

ambv committed:
Work around mypy being unhappy with _colorize

ambv committed:
Remove duplicate test

ambv committed:
Add types to HistoricalReader and CompletingReader

JelleZijlstra commented on Lib/_pyrepl/_minimal_curses.py:
```diff
@@ -0,0 +1,68 @@
+        path = ctypes.util.find_library(lib)
+        if path:
+            return path
+    raise ModuleNotFoundError("curses library not found", name="_minimal_curses")
```
```suggestion
    raise ModuleNotFoundError("curses library not found", name="_pyrepl._minimal_curses")
```

JelleZijlstra commented on Lib/_pyrepl/__main__.py:
```diff
@@ -0,0 +1,42 @@
+    run_interactive = None
+    try:
+        if not os.isatty(sys.stdin.fileno()):
+            raise ImportError
```
Maybe raise a more specific exception here?

JelleZijlstra commented on Lib/traceback.py:
```diff
@@ -140,7 +140,7 @@ def _print_exception_bltin(exc, /):
 
 def format_exception(exc, /, value=_sentinel, tb=_sentinel, limit=None, \
-                     chain=True):
+                     chain=True, **kwargs):
```
Why doesn't this work?
```suggestion
                     chain=True, colorize=False):
```
Adding `**kwargs` to the public signature would be unfortunate.

Could also make it a private parameter, if you don't want to document it:
```suggestion
                     chain=True, _colorize=False):
```

pablogsal replied:
> Why doesn't this work?

I don't want to be exposed in the signature even as `_colorize`. That is how I approached the general colorisation.

ambv committed:
Add typing to readline.py

ambv committed:
Pin test_pyrepl to the `curses` test resource

pablogsal committed:
Don't call anything curses in test_repl

ambv committed:
Many machines on CI run with -uall; skip the test if curses import fails

pablogsal committed:
Fix refleaks

ambv committed:
Document PYTHON_BASIC_REPL

pablogsal committed:
Do not call _setup() on import

ambv committed:
Document pyrepl

ambv committed:
Merge branch 'main' into pyrepl

hugovk commented on Doc/tutorial/appendix.rst:
```diff
@@ -10,6 +10,28 @@ Appendix
+This one supports color, multiline editing, history browsing, and
+paste mode.  To disable color, see :ref:`using-on-controlling-color` for
+details.  Function keys provide some additional functionality.
+``F1`` enters the interactive help browser :mod:`pydoc`.
```
There's a Sphinx role for keyboard entry:

```suggestion
:kbd:`F1` enters the interactive help browser :mod:`pydoc`.
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-kbd

hugovk commented on Doc/tutorial/appendix.rst:
```diff
@@ -10,6 +10,28 @@ Appendix
+paste mode.  To disable color, see :ref:`using-on-controlling-color` for
+details.  Function keys provide some additional functionality.
+``F1`` enters the interactive help browser :mod:`pydoc`.
+``F2`` allows for browsing command-line history without output nor the
```
```suggestion
:kbd:`F2` allows for browsing command-line history without output nor the
```

hugovk commented on Doc/tutorial/appendix.rst:
```diff
@@ -10,6 +10,28 @@ Appendix
+details.  Function keys provide some additional functionality.
+``F1`` enters the interactive help browser :mod:`pydoc`.
+``F2`` allows for browsing command-line history without output nor the
+:term:`>>>` and :term:`...` prompts. ``F3`` enters "paste mode", which
```
```suggestion
:term:`>>>` and :term:`...` prompts. :kbd:`F3` enters "paste mode", which
```

hugovk commented on Doc/tutorial/appendix.rst:
```diff
@@ -10,6 +10,28 @@ Appendix
+``F1`` enters the interactive help browser :mod:`pydoc`.
+``F2`` allows for browsing command-line history without output nor the
+:term:`>>>` and :term:`...` prompts. ``F3`` enters "paste mode", which
+makes pasting larger blocks of code easier. Press ``F3`` to return to
```
```suggestion
makes pasting larger blocks of code easier. Press :kbd:`F3` to return to
```

hugovk commented on Doc/tutorial/appendix.rst:
```diff
@@ -10,6 +10,28 @@ Appendix
+makes pasting larger blocks of code easier. Press ``F3`` to return to
+the regular prompt.
+
+When using the new interactive shell, exit the shell by typing ``exit``
```
```suggestion
When using the new interactive shell, exit the shell by typing :kbd:`exit`
```

hugovk commented on Doc/tutorial/appendix.rst:
```diff
@@ -10,6 +10,28 @@ Appendix
+the regular prompt.
+
+When using the new interactive shell, exit the shell by typing ``exit``
+or ``quit``. Adding call parentheses after those commands is not
```
```suggestion
or :kbd:`quit`. Adding call parentheses after those commands is not
```

hugovk reviewed (commented):
Please also add a What's New entry.

pablogsal committed:
Apply suggestions from code review

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

pablogsal committed:
Update Lib/_pyrepl/_minimal_curses.py

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

pablogsal committed:
Use more specific exceptions

hugovk reviewed (approved):
Docs changes look good.

One suggestion: this PR adds the main docs to the tutorial, we should probably add something to the reference, but this could be a followup. 

> > Please also add a What's New entry.
> 
> There is already one (Misc/NEWS.d/next/Core and Builtins/2024-04-28-00-41-17.gh-issue-111201.cQsh5U.rst)

That's the changelog, this change definitely needs highlighting in https://docs.python.org/3.13/whatsnew/3.13.html but it can also be in a followup PR.

ambv committed:
Use better error messages

ambv committed:
Add to What's New

ambv committed:
Forgot the PyPy attribution in "What's New"

cfbolz commented on Lib/_pyrepl/reader.py:
```diff
@@ -0,0 +1,660 @@
+            self.restore()
+            yield
+        finally:
+            for arg in ("msg", "ps1", "ps2", "ps3", "ps4", "paste_mode"):
```
why is there a discrepancy between how `prev_state` is constructed (which uses `fields(self)` and how it is used in the finally block (listing the attributes explicitly)?

ambv replied:
It evolved into this, which you're right would now be cleaner to just reuse the list of fields.

I originally started with just `dict(__dict__)` but when I added typing with `slots=True` the `__dict__` was gone so I modified it to use `fields()` instead.

cfbolz reviewed (approved):
Cool to see this!

Just to check, was it intentional that you didn't port [PyPy's `_pyrepl` tests](https://github.com/pypy/pypy/tree/branches/py3.10/extra_tests/test_pyrepl)? There's not a huge amount of tests there, admittedly.

I am philosophically a little bit worried that people will start using and relying on `_pyrepl` internal details, which are now slightly different between PyPy and CPython, right? This could lead to PyPy's life getting somewhat harder in this area. As usual, it's hard to stop people from doing that, of course.

hugovk commented on Doc/whatsnew/3.13.rst:
```diff
@@ -97,6 +97,34 @@ New typing features:
+
+For more on interactive mode, see :ref:`tut-interac`.
+
+(Contributed by Pablo Galindo Galgado, Łukasz Langa, and
```
```suggestion
(Contributed by Pablo Galindo Salgado, Łukasz Langa, and
```

ambv committed:
let's never talk about this

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>
